File: next-js-boilerplate/src/validators/forms/advanced.ts
import { z } from "zod";

export function createAdvancedSchemas(t: Record<string, string>) {
  const memberName = z
    .string()
    .min(1, t.memberNameRequired ?? "Name is required");
  const memberEmail = z.string().email(t.memberEmailInvalid ?? "Invalid email");
  const memberRole = z.string().min(1, t.memberRoleRequired ?? "Select a role");

  const memberSchema = z.object({
    name: memberName,
    email: memberEmail,
    role: memberRole,
  });

  return {
    fullName: z.string().min(2, t.fullNameMin ?? "At least 2 characters"),
    email: z.string().email(t.emailInvalid ?? "Invalid email"),
    password: z.string().min(8, t.passwordMin ?? "At least 8 characters"),
    companyName: z
      .string()
      .min(1, t.companyNameRequired ?? "Company name is required"),
    taxId: z
      .string()
      .regex(
        /^[A-Z]{2}\d{2,13}[A-Z0-9]$/,
        t.taxIdInvalid ?? "Invalid tax ID format",
      ),
    industry: z.string().min(1, t.industryRequired ?? "Select an industry"),
    members: z.array(memberSchema).min(0),
    memberName,
    memberEmail,
    memberRole,
  };
}

File: flutter-boilerplate/lib/validators/forms/advanced.dart
String? validateFullName(String? value) {
  if (value == null || value.isEmpty) return 'Name is required';
  if (value.length < 2) return 'At least 2 characters';
  return null;
}

String? validateEmail(String? value) {
  if (value == null || value.isEmpty) return 'Email is required';
  final emailRegex = RegExp(r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$');
  if (!emailRegex.hasMatch(value)) return 'Invalid email';
  return null;
}

String? validatePassword(String? value) {
  if (value == null || value.isEmpty) return 'Password is required';
  if (value.length < 8) return 'At least 8 characters';
  return null;
}

String? validateCompanyName(String? value) {
  if (value == null || value.isEmpty) return 'Company name is required';
  return null;
}

String? validateTaxId(String? value) {
  if (value == null || value.isEmpty) return null;
  final taxRegex = RegExp(r'^[A-Z]{2}\d{2,13}[A-Z0-9]$');
  if (!taxRegex.hasMatch(value)) return 'Invalid tax ID format';
  return null;
}

String? validateIndustry(String? value) {
  if (value == null || value.isEmpty) return 'Select an industry';
  return null;
}

String? validateMemberName(String? value) {
  if (value == null || value.isEmpty) return 'Name is required';
  return null;
}

String? validateMemberEmail(String? value) {
  if (value == null || value.isEmpty) return 'Email is required';
  final emailRegex = RegExp(r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$');
  if (!emailRegex.hasMatch(value)) return 'Invalid email';
  return null;
}

String? validateRole(String? value) {
  if (value == null || value.isEmpty) return 'Select a role';
  return null;
}
